src/accountsdb/db.zig

```diff
@@ -1072,24 +1148,67 @@ pub const AccountsDB = struct {
                 }
 
                 // flush the slots
+                try unclean_account_files.ensureTotalCapacityPrecise(flush_slots.items.len);
                 for (flush_slots.items) |flush_slot| {
-                    self.flushSlot(flush_slot, &unclean_account_files) catch |err| {
+                    const unclean_file_id = self.flushSlot(flush_slot) catch |err| {
                         // flush fail = loss of account data on slot -- should never happen
                         self.logger.errf("flushing slot {d} error: {s}", .{ flush_slot, @errorName(err) });
+                        continue;
                     };
+                    unclean_account_files.appendAssumeCapacity(unclean_file_id);
                 }
+                self.largest_flushed_slot.store(root_slot, .seq_cst);
+            }
+
+            const largest_flushed_slot = self.largest_flushed_slot.load(.seq_cst);
+            const latest_full_snapshot_slot = blk: {
+                const latest_full_snapshot_info, var latest_full_snapshot_info_lg = self.latest_full_snapshot_info.writeWithLock();
+                defer latest_full_snapshot_info_lg.unlock();
+                break :blk if (latest_full_snapshot_info.*) |info| info.slot else 0;
+            };
+
+            if (largest_flushed_slot - latest_full_snapshot_slot >= slots_per_full_snapshot) {
+                var snapshot_gen_pkg, const snapshot_gen_info = try self.makeFullSnapshotGenerationPackage(
+                    largest_flushed_slot,
+                    random_bank_fields,
+                    rand.random().int(u64),
+                    0,
+                );
+                defer snapshot_gen_pkg.deinit();
+
+                const archive_file_name_bounded = sig.accounts_db.snapshots.FullSnapshotFileInfo.snapshotNameStr(.{
+                    .slot = snapshot_gen_info.slot,
+                    .hash = snapshot_gen_info.hash,
+                    .compression = .zstd,
+                });
+                const archive_file_name = archive_file_name_bounded.constSlice();
+                const archive_file = try self.snapshot_dir.createFile(archive_file_name, .{ .read = true });
+                defer archive_file.close();
+
+                const zstd_write_ctx = zstd.writerCtx(archive_file.writer(), &zstd_compressor, zstd_buffer);
+
+                try snapshot_gen_pkg.write(zstd_write_ctx.writer());
+                try zstd_write_ctx.finish();
+
+                try self.commitFullSnapshotInfo(snapshot_gen_info, .delete_old);
```

Comment:
would you want to put this logic into a method? seems its used both here and in tests

Reply:
The requirements of the code in the unit test and here are subtly different; in the unit test, we retain the archive file, whilst here, we want to close it as soon as we're finished writing the snapshot.